R/civis_ml_utils.R

```diff
@@ -19,19 +19,93 @@ CIVIS_ML_CLASSIFIERS <- c("sparse_logistic",
+                                       object_type='template_script',
+                                       function(x) length(x) == 0)
+
+
+  template_aliases <- sapply(template_alias_objects, `[[`, 'alias')
+
+  # Template script aliases are not used exclusively for CivisML
+  civisml_template_alias_objects <- template_alias_objects[startsWith(template_aliases,
+                                                                      'civis-civisml-')]
+
+  ids <- lapply(civisml_template_alias_objects, function(alias_obj){
+
+        id = alias_obj$objectId
+
+        job_type_version = get_job_type_version(alias_obj$alias)
+
+        return(list(id = id,
+                    version = job_type_version$version,
+                    name = job_type_version$job_type))
+
+    })
+
+  CIVIS_ML_TEMPLATE_IDS <- data.frame(matrix(unlist(ids),
+                                  nrow=length(ids),
+                                  byrow=T),
+                           stringsAsFactors=FALSE)
+
+  names(CIVIS_ML_TEMPLATE_IDS) <- c("id", "version", "name")
+  CIVIS_ML_TEMPLATE_IDS[,"id"] <- as.integer(CIVIS_ML_TEMPLATE_IDS[,"id"])
+
+
+  return(CIVIS_ML_TEMPLATE_IDS)
+
+}
+
+get_job_type_version <- function(alias){
+  # Derive the job type and version from the given alias.
+
+  # `alias` is a character vector and length(`alias`) = 1
+
+  # A version-less alias for production, e.g., "civis-civisml-training"
+  match_production <- unlist(regmatches(alias, regexec('\\Acivis-civisml-(\\w+)\\Z', alias, perl=TRUE)))
+  # A versioned alias, e.g., "civis-civisml-training-v2-3"
+  match_v <- unlist(regmatches(alias, regexec('\\Acivis-civisml-(\\w+)-v(\\d+)-(\\d+)\\Z', alias, perl=TRUE)))
+  # A special-version alias, e.g., "civis-civisml-training-dev"
+  match_special <- unlist(regmatches(alias, regexec('\\Acivis-civisml-(\\w+)-(\\S+[^-])\\Z', alias, perl=TRUE)))
+
+
+  if (!identical(match_production, character(0))) {
+
+    job_type = match_production[2]
+    version = "prod"
+
+  } else if (!identical(match_v, character(0))) {
+
+    job_type = match_v[2]
+    version = paste0("v",match_v[3],".",match_v[4])
+
+  } else if (!identical(match_special, character(0))) {
+
+    job_type = match_special[2]
+    version = match_special[3]
+
+  } else {
+
+    stop(paste('Unable to parse the job type and version from the CivisML alias:', alias))
+
+  }
+
+  return(list(job_type = job_type, version = version))
+
+}
 
 # returns a version compatible template id for a given training model without API calls.
 get_predict_template_id <- function(m) {
   train_id <- m$job$fromTemplateId
   this_version <- CIVIS_ML_TEMPLATE_IDS[CIVIS_ML_TEMPLATE_IDS$id == train_id, "version"]
-  CIVIS_ML_TEMPLATE_IDS[CIVIS_ML_TEMPLATE_IDS$version == this_version &
-                        CIVIS_ML_TEMPLATE_IDS$name == "predict", "id"]
+
+  #  `prod` points to the same template id as the latest version (e.g. v2.2)
+  CIVIS_ML_TEMPLATE_IDS[CIVIS_ML_TEMPLATE_IDS$version == this_version[1] &
```

Comment:
Is this too kludgey to address the fact that `prod` points to the same template id as the latest version (e.g. v2.2) so that when `train_id` corresponds to the `prod` template id `this_version` will be a vector equal to e.g. `"prod", "v2.2"`)

Reply:
Grabbing `prod` by name is clearer to read I think.

Reply:
Sorry I'm not sure if I fully understand your comment..

Stepping through this function 

 `train_id <- m$job$fromTemplateId` is grabbed from the argument `m`

`civis_ml_template_ids <- get_template_ids_all_versions()` produces

```
civis_ml_template_ids
      id version         name
1  11219    prod     training
2  11220    prod   prediction
3  11221    prod registration
4  10615     dev     training
5  10616     dev   prediction
6  11028     dev registration
7  11219    v2.2     training
8  11220    v2.2   prediction
9  11221    v2.2 registration
10  7020    v0.5     training
11  7021    v0.5   prediction
12  8387    v1.0     training
13  8388    v1.0   prediction
14  9112    v1.1     training
15  9113    v1.1   prediction
16  9968    v2.0     training
17  9969    v2.0   prediction
18 10582    v2.1     training
19 10583    v2.1   prediction
```

If `train_id = 11219` then 

`this_version <- civis_ml_template_ids[civis_ml_template_ids$id == train_id, "version"]` produces 

`"prod" "v2.2"`

Is your suggestion to just return the `prediction` template id for `prod` and not based off of the the argument `m` . Would we have to refactor this when we make changes to make the version run a parameter?

Reply:
Okay. I kept it as indexing by integer! Thanks